### ISSUE TITLE
collectd python plugin failing

### ISSUE THREAD

**Issue by dpb587** [logsearch/logsearch-boshrelease#26]
The python plugin (used by several plugins here) is causing collectd unhappiness...

```
ERROR: lt_dlopen ("/var/vcap/packages/collectd/lib/collectd/python.so") failed: file not found. The most common cause for this problem are missing dependencies. Use ldd(1) to check the dependencies of the plugin / shared object.
```

This may partially be an artifact of switching to the newer stemcells.

**Comment by mrdavidlaing** [comment]
> This may partially be an artefact of switching to the newer stemcells.

I'm pretty sure I saw chatter on the BOSH mailing list saying that newer stemcells will ship with _less_ installed by default.  This might be an artefact of that.

**Comment by dpb587** [comment]
@mrdavidlaing, is there a more proper way to installing a dependency on job vms than to add something like this in the control script before starting?

```
[[ -f /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0 ]] || apt-get install -y libpython2.7
```

**Comment by mrdavidlaing** [comment]
Yes - you should have a separate package that bundles the required files into the bosh-release

The issue with the above is that it won't work on servers with no Internet connection

**Comment by dpb587** [comment]
Please point me to an example of how another bosh release does this? The `libpython2.7.so.1.0` depends on several other system modules itself - it seems extremely unfortunate to be hard-coding this sort of dependency outside of real package managers.

**Comment by mrdavidlaing** [comment]
Ideally we could find another boshrelease which already has a python package, and just crib that.

Otherwise, bosh-gen allows you to make packages from apt-get sources - https://github.com/cloudfoundry-community/bosh-gen#fast-way---reuse-aptitudedebian-packages

**Comment by mrdavidlaing** [comment]
Aha! - https://github.com/murraju/insightfactory-boshrelease/tree/master/packages/python

**Comment by mrdavidlaing** [comment]
Or - https://github.com/drnic/gitlabhq-boshrelease/tree/master/packages/python27

A Github search for `boshrelease python` across all repos found these

**Comment by sopel** [comment]
Because it's such an obvious example, I'd like to mention again that I consider this BOSH approach to package management to be detrimental to the proclaimed goal of being in control of your data center stability and security:

While it is completely sound to insist on having the packages to install to be available locally _and_ as source code in order to achieve complete upstream and internet independence, it seems not only extremely unfortunate, but actually dangerous to make people run around and hunt for some random and usually slightly outdated boshrelease or do it themselves rather than rely on the peer reviewed and security monitored upstream packaging of established and renowned projects like Debian.

From that angle I really appreciate Dr. Nic's [bosh-gen](https://github.com/cloudfoundry-community/bosh-gen#fast-way---reuse-aptitudedebian-packages) initiative, but remain confused that the BOSH community doesn't seem to pick this up and promote it as a better alternative.

**Comment by sopel** [comment]
I said to: 

> [...] remain confused that the BOSH community doesn't seem to pick this up and promote it as a better  alternative.

Well, there's actually a reasonable explanation in that very section about [reusing Debian packages](https://github.com/cloudfoundry-community/bosh-gen#fast-way---reuse-aptitudedebian-packages):

> This is the last option, and it is not the best option. Many Debian packages will also start processes that have default configuration that is not correct for your use case. It may be fast to get the Debian packages; but additional work may be required by your jobs to stop and unhook the processes that are automatically started upon installation.

**Comment by aristotelesneto** [comment]
Just to chime in, as James Bayer said on the [Pivotal blog](http://blog.gopivotal.com/cloud-foundry-pivotal/products/remote-dependencies-convenience-risk-and-other-considerations-for-operating-distributed-systems), theres no right or wrong answer - simply tradeoffs to consider.

One of the reasons why I have personally preferred to compile the dependencies as BOSH packages, is that I'm then responsible for the versions of the system as a whole - and once I know it works, I know that if I need to rebuild it from scratch it will behave exactly the same way as it did before - as opposed to something misbehaving as a result of a new version bundled within the stemcell.

**Comment by sopel** [comment]
Thanks for chiming in, that's a helpful cross reference indeed - I do not really buy the _externally hosted package manager_ argument, because you could store Debian or CentOS packages locally just as easily as the source archive, however, the cross distribution argument is a valid one in general, and all the more important for a deliberately and highly portable solution like BOSH.

**Comment by mrdavidlaing** [comment]
This has been addressed by #27
